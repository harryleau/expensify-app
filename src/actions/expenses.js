import database from '../firebase/firebase';

//////// NO FIREBASE 
// component calls action generator
// action generator returns object
// component calls action generator
// redux store changes

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

//////// WITH FIREBASE
// component calls action generator
// action generator returns function
// component dispatches function - startAddExpense returns a function that dispatches action addExpense
// function runs (has the ability to dispatch other actions and do whatever it wants)

// START_ADD_EXPENSE
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
    const expense = { description, note, amount, createdAt };

    // for further promise chaining, we use return. for example, in test file, we can chain assertion after calling startAddExpense
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});