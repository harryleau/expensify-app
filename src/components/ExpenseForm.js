import React from 'react';

export default class ExpenseForm extends React.Component {
  // new syntax for constructor
  state = {
    description: '',
    note: ''
  };

  // Solution 1: set value to a const.
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  // Solution 2: use e.persist(), cannot use e.target.value in a callback without e.persist()
  onNoteChange = e => {
    e.persist();
    this.setState(() => ({ note: e.target.value }));
  }

  onAmountChange = e => {
    const amount = e.target.value;
    if(amount.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  render() {
    return (
      <div>
        <form>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <textarea 
            placeholder="Add a note for your expense"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
};