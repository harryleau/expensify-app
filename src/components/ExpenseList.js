import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from './../selectors/expenses';

export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>

    <div className="list-body">
      {
        props.expenses.length === 0 ? (
          <div>
            <span className="list-item list-item--message">No expenses</span>
          </div>
        ) : (
          props.expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
        )
      }
    </div>
  </div>
);

const mapStatetoProps = (state) => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStatetoProps)(ExpenseList);

// the connect method will return a function like requestAuthentication in hoc.js so we call that function right after that with the regular component ExpenseList
// connect also takes in a function that takes in state and return a props obj that can be used in ExpenseList component. that function argument is famously named mapStateToProps. 
// the final default export will be a higher order component of ExpenseList.