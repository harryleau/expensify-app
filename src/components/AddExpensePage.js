import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from './../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

// just like mapStateToProps, it turns dispatch actions into props, so in the component we don't need to use props.dispatch(addExpense(expense)) anymore.
const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
});

// first arg is mapStateToProps, we don't need it in this component, so set it to be undefined
export default connect(undefined, mapDispatchToProps)(AddExpensePage);