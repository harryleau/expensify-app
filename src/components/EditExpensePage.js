import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from './../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (updatedExpense) => {
    this.props.editExpense(this.props.expense.id, updatedExpense);
    this.props.history.push('/');
  };

  onRemove = () => {
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
}

// not only we can access state, but we can also access current props through second argument to use some props like match.params
// find() is an array method that will return the value if the condition is true
const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (id) => dispatch(removeExpense({ id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);