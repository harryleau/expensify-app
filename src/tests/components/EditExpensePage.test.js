import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from './../../components/EditExpensePage';
import expenses from './../fixtures/expenses';

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<EditExpensePage expense={expenses[2]} editExpense={editExpense} removeExpense={removeExpense} history={history} />);
});

test('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {
  const updatedExpense = { note: 'updated' }
  wrapper.find('ExpenseForm').prop('onSubmit')(updatedExpense);
  expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, updatedExpense);
  expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should handle removeExpense', () => {
  wrapper.find('button').simulate('click');
  expect(removeExpense).toHaveBeenLastCalledWith(expenses[2].id);
  expect(history.push).toHaveBeenLastCalledWith('/');
});