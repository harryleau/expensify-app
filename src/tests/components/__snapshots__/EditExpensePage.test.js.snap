// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 3599,
        "createdAt": 259200000,
        "description": "Games",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
